contracts/bridges/BaseBridgeReceiver.sol

```diff
@@ -0,0 +1,92 @@
+
+    event Initialized(address indexed l2Timelock, address indexed mainnetTimelock);
+    event NewL2Timelock(address indexed oldL2Timelock, address indexed newL2Timelock);
+    event NewMainnetTimelock(address indexed oldMainnetTimelock, address indexed newMainnetTimelock);
+    event ProcessMessage(address indexed messageSender, address[] targets, uint256[] values, string[] signatures, bytes[] calldatas, uint eta);
+
+    address public initializer;
+    address public mainnetTimelock;
+    address public l2Timelock;
+
+    constructor(address _initializer) {
+        initializer = _initializer;
+    }
+
+    function initialize(address _mainnetTimelock, address _l2Timelock) external {
+        if (initializer == address(0)) revert AlreadyInitialized();
+        if (msg.sender != initializer) revert Unauthorized();
+        mainnetTimelock = _mainnetTimelock;
+        l2Timelock = _l2Timelock;
+        initializer = address(0);
+        emit Initialized(_mainnetTimelock, _l2Timelock);
+    }
+
+    function acceptL2TimelockAdmin() external {
+        if (msg.sender != l2Timelock) revert Unauthorized();
+        ITimelock(l2Timelock).acceptAdmin();
+    }
+
+    function setL2Timelock(address newTimelock) public {
+        if (msg.sender != l2Timelock) revert Unauthorized();
+        address oldL2Timelock = l2Timelock;
+        l2Timelock = newTimelock;
+        emit NewL2Timelock(oldL2Timelock, newTimelock);
+    }
+
+    function setMainnetTimelock(address newTimelock) public {
+        if (msg.sender != l2Timelock) revert Unauthorized();
+        address oldMainnetTimelock = mainnetTimelock;
+        mainnetTimelock = newTimelock;
+        emit NewMainnetTimelock(oldMainnetTimelock, newTimelock);
+    }
+
+    function processMessage(
+        address messageSender,
+        bytes calldata data
+    ) internal {
+        if (messageSender != mainnetTimelock) revert Unauthorized();
+
+        address[] memory targets;
+        uint256[] memory values;
+        string[] memory signatures;
+        bytes[] memory calldatas;
+
+        (targets, values, signatures, calldatas) = abi.decode(
+            data,
+            (address[], uint256[], string[], bytes[])
+        );
+
+        if (values.length != targets.length) revert BadData();
+        if (signatures.length != targets.length) revert BadData();
+        if (calldatas.length != targets.length) revert BadData();
+
+        uint delay = ITimelock(l2Timelock).delay();
+        uint eta = block.timestamp + delay;
+
+        for (uint8 i = 0; i < targets.length; ) {
+            ITimelock(l2Timelock).queueTransaction(targets[i], values[i], signatures[i], calldatas[i], eta);
+            unchecked { i++; }
+        }
+
+        emit ProcessMessage(messageSender, targets, values, signatures, calldatas, eta);
+    }
+
+    function executeTransaction(
+        address target,
+        uint value,
+        string memory signature,
+        bytes memory data,
+        uint eta
```

Comment:
I still think that assigning each action an id and using that to execute is way easier, especially since each action needs to be executed separately and there could be 5+ actions per proposal just for updating supply caps.

Reply:
Assigning action ids will also allow us to more easily support action states (akin to proposal states in GovernorBravo) such as queued, executed, expired. Not sure if we really want to have action states here, but it would better align with the interfaces we already use in `ProposalConstraint`.

Reply:
synced offline with Kevin.

we realized that we really need this to be storing the equivalent to the Proposals on GovernorBravo: a set of transactions that should be executed atomically

- [ ] implement Proposals for BaseBridgeReceiver

Reply:
Discussed this offline with Scott. He brought up a good point that we could actually store these actions as proposals instead. That way, proposals keep their atomic, all-or-nothing property. It was kind of an oversight that we initially allowed different actions from the same proposal to be executed individually (selectively and in any order).

Reply:
updated; `exectureTransaction(/.. ../)` is now `executeProposal(proposalId)`

added Proposal struct type, a `state(proposalId)` function, Proposal events and Proposal errors.